### PR TITLE
Proxied TLS passthrough

### LINKED ISSUE #3758: proxied-tls behind a TCP loadbalancer with proxy protocol
## Description of the problem
The `proxied-tls` transport for network source only supports reading the PROXY header inside the encrypted payload.
In practice, when using haproxy for example in front of a syslog-ng server, the PROXY header will be outside the TLS payload as the TLS termination is performed on syslog-ng side.
When using `proxied-tls` with the previous configuration, syslog-ng raises an SSL error as the TCP packet contains the PROXY header first , but syslog-ng is expecting a TLS handshake client hello first

To be honest, except when using `loggen -U -H` for tests, I'm not aware of a way of having the PROXY header inside the encrypted payload as the remote logging clients are not the ones adding the PROXY header

## Proposed solution
A possible solution could be support reading the PROXY header before performing the TLS handshake

## Alternatives
The current workaround I could find when using haproxy and performing TLS termination on the syslog-ng machine, is having locally another haproxy installed on the syslog machine configured to accept proxy protocol and perform the TLS offload, then forward the decrypted data and the initial PROXY header to syslog-ng which will be using `proxied-tcp`

If I'm mistaken about `proxied-tls` option, or the use case is not clear, I can give more context of course
Thanks

### PR DESCRIPTION
**Pull request by OverOrion**
This PR adds a new transport method `proxied-tls-passthrough` which is capable of detecting the PROXY header before the TLS payload. Loggen has been updated with the ` -proxied-passthrough ` option for testing purposes.

TODO:
~~- [x] add news file~~

Resolves #3758

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by mitzkia** [comment]
@kira-syslogng test this please test branch=exclude_https_testcases_from_run_local;

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/logproto/logproto-proxied-text-server.c [hunk 1]
```diff
@@ -189,6 +189,34 @@ _log_proto_proxied_text_server_add_aux_data(LogProtoProxiedTextServer *self, Log
   return;
 }
 
+static inline LogProtoStatus
+_fetch_into_proxy_buffer(LogProtoProxiedTextServer *self, gsize *hdr_len)
+{
+  for(gint i = 0; i < PROXY_PROTO_HDR_MAX_LEN_RFC; i++)
+    {
+      gssize rc = log_transport_read(self->super.super.super.transport, &(self->v1_proxy_header_buff[i]), sizeof(gchar),
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply lib/logproto/logproto-proxied-text-server.c:197]: Protos are non-blocking, `EAGAIN` has to be handled properly. which means that the for loop (in the current form) won't work either, because `_fetch_into_proxy_buffer()` will be called again in that case, and we should be able to continue reading the header where we left off.

When detecting `EAGAIN`, `LPS_AGAIN` must be returned from the proto implementation.
We also need to implement the `prepare()` virtual method, which should the IO direction to `G_IO_IN` and return `LPPA_POLL_IO` when we know that an EAGAIN was previously detected during reading the proxy header. Otherwise, we should let the parent class' prepare method to decide about the next step (by publishing and calling `log_proto_text_server_prepare()`).

File: modules/afsocket/transport-mapper-inet.c [hunk 2]
```diff
@@ -391,12 +391,13 @@ transport_mapper_network_apply_transport(TransportMapper *s, GlobalConfig *cfg)
       self->super.sock_proto = IPPROTO_TCP;
       self->require_tls = TRUE;
     }
-    else if (strcasecmp(transport, "proxied-tls-passthrough") == 0)
+  else if (strcasecmp(transport, "proxied-tls-passthrough") == 0)
```
Comment (MrAnno) [inline modules/afsocket/transport-mapper-inet.c:394]: This does not belong to this commit. Please squash and/or fix your commits.

**Review by MrAnno — CHANGES_REQUESTED** [review]

File: lib/logproto/logproto-proxied-text-server.c [hunk 3]
```diff
@@ -189,6 +189,41 @@ _log_proto_proxied_text_server_add_aux_data(LogProtoProxiedTextServer *self, Log
...
+          self->v1_proxy_header_buff[i+1] = '\0';
+          *hdr_len = i+1;
+          if (self->v1_proxy_header_buff[i] == '\n')
+            {
+              return LPS_SUCCESS;
+            }
+        }
+      else if(rc == 0)
+        {
+          return LPS_AGAIN;
```
Comment (MrAnno) [inline lib/logproto/logproto-proxied-text-server.c:210]: This is an EOF situation, which usually means `LPS_EOF`.

LPS_AGAIN can be detected with `rc < 0 && errno == EAGAIN`

File: lib/logproto/logproto-proxied-text-server.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrAnno) [reply lib/logproto/logproto-proxied-text-server.c:197]: Note: The transport's IO condition has priority over the proto's, so it has to be taken into account as well (see other proto implementations):

```
  *cond = self->super.transport->cond;

  /* if there's no pending I/O in the transport layer, then ... */
  if (*cond == 0)
```

File: lib/logproto/logproto-proxied-text-server.c [hunk 4]
```diff
@@ -201,9 +229,16 @@ _log_proto_proxied_text_server_handshake(LogProtoServer *s)
   LogProtoProxiedTextServer *self = (LogProtoProxiedTextServer *) s;
 
   // Fetch a line from the transport layer
-  status = log_proto_buffered_server_fetch(&self->super.super.super, &msg, &msg_len, &may_read, NULL, NULL);
-
-  self->handshake_done = (status == LPS_SUCCESS);
+  if(self->has_to_switch_to_tls)
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply lib/logproto/logproto-proxied-text-server.c:239]: `v1_proxy_header_buff` is completely unused in case STARTTLS is not required.

We could completely remove this conditional and always use the new proxy header fetch method.
There might be 2 reasons for not using the unbuffered fetch method: performance and the fear of regression.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by mitzkia** [comment]
I like the Light commits and of course thanks for that.
Could you please also add 1 or 2 negative testcases related to 'proxied-tls-passthrough', with miss configuring loggen or transport option for network source driver against incoming message.

File: modules/afsocket/transport-mapper-inet.c [hunk 5]
```diff
@@ -144,7 +144,7 @@ transport_mapper_inet_construct_log_transport(TransportMapper *s, gint fd)
 {
   TransportMapperInet *self = (TransportMapperInet *) s;
 
-  if (self->tls_context && _is_tls_required(self))
+  if (self->tls_context && _is_tls_required(self) && !self->super.create_multitransport)
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/afsocket/transport-mapper-inet.c:147]: We've found a way to completely get rid of these boolean tricks.

**Comment by OverOrion** [comment]
@mitzkia 
Thank you for the suggestion! I am going to write those tests, but if it is okay with you that would be in a separate PR. :)
Also kudos to @szemere and @MrAnno they greatly helped me with this PR!

File: lib/logproto/logproto-proxied-text-server.c [hunk 6]
```diff
@@ -214,6 +249,12 @@ _log_proto_proxied_text_server_handshake(LogProtoServer *s)
   if (parsable)
     {
       msg_info("PROXY protocol header parsed successfully");
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply lib/logproto/logproto-proxied-text-server.c:281]: We could not do this properly, because LogProto instances currently have no LogPipe/Driver/location information.

`evt_tag_printf("driver", "%p", self)` has no real meaning, it's actually misleading, because we print the LogProto address instead of the driver.

File: lib/logproto/logproto-proxied-text-server.c [hunk 7]
```diff
@@ -189,31 +189,100 @@ _log_proto_proxied_text_server_add_aux_data(LogProtoProxiedTextServer *self, Log
...
+static void
+_log_proto_proxied_text_server_switch_to_tls(LogProtoProxiedTextServer *self)
+{
+  if (!multitransport_switch((MultiTransport *)self->super.super.super.transport, transport_factory_tls_id()))
+    {
+      msg_error("proxied-tls failed to switch to TLS");
+      return;
+    }
+  msg_info("proxied-tls switch to TLS: OK");
+}
```
Comment (MrAnno) [inline lib/logproto/logproto-proxied-text-server.c:260]: Please move this to debug level (it was there previously, but it got reverted accidentally).

File: lib/logproto/logproto-proxied-text-server.c [hunk 8]
```diff
@@ -189,31 +189,100 @@ _log_proto_proxied_text_server_add_aux_data(LogProtoProxiedTextServer *self, Log
...
-  msg_debug("PROXY protocol header received", evt_tag_printf("line", "%.*s", (gint) msg_len, msg));
+  msg_debug("PROXY protocol header received", evt_tag_printf("line", "%.*s",
+                                                             (gint) self->proxy_header_buff_len, self->proxy_header_buff));
 
   if (parsable)
     {
       msg_info("PROXY protocol header parsed successfully");
+
+      if (self->has_to_switch_to_tls)
+        _log_proto_proxied_text_server_switch_to_tls(self);
```
Comment (MrAnno) [inline lib/logproto/logproto-proxied-text-server.c:284]: I think we should close the connection if we fail to switch to TLS, otherwise, we'd start processing encrypted payload without decrypting it. syslog-ng handles almost everything in no-parse mode, so it could be a real problem :)

For example:
```
if (self->has_to_switch_to_tls && !_log_proto_proxied_text_server_switch_to_tls(self))
  return LPS_ERROR;
```

**Review by MrAnno — APPROVED** [review]
LGTM. I found one more thing that I think we should fix.

I've triple-checked the logproto implementation, but I was not thorough with the review of the loggen modification.

**Review by MrAnno — DISMISSED** [review]

**Comment by kira-syslogng** [comment]
Build FAILURE

File: lib/logproto/logproto-proxied-text-server.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kokan) [reply lib/logproto/logproto-proxied-text-server.c:283]: At the very least this should not be on info level.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — APPROVED** [review]
Thanks! :)
